das/google_cloud_storage_service.go

```diff
@@ -0,0 +1,191 @@
+		operator:     &GoogleCloudStorageClient{client: client},
+		bucket:       config.Bucket,
+		objectPrefix: config.ObjectPrefix,
+		enableExpiry: config.EnableExpiry,
+		maxRetention: config.MaxRetention,
+	}, nil
+}
+
+func (gcs *GoogleCloudStorageService) Put(ctx context.Context, value []byte, timeout uint64) error {
+	logPut("das.GoogleCloudStorageService.Store", value, timeout, gcs)
+	if err := gcs.operator.Upload(ctx, gcs.bucket, gcs.objectPrefix, value); err != nil {
+		log.Error("das.GoogleCloudStorageService.Store", "err", err)
+		return err
+	}
+	return nil
+}
+
+func (gcs *GoogleCloudStorageService) GetByHash(ctx context.Context, key common.Hash) ([]byte, error) {
+	log.Trace("das.GoogleCloudStorageService.GetByHash", "key", pretty.PrettyHash(key), "this", gcs)
+	buf, err := gcs.operator.Download(ctx, gcs.bucket, gcs.objectPrefix, key)
+	if err != nil {
+		log.Error("das.GoogleCloudStorageService.GetByHash", "err", err)
+		return nil, err
+	}
+	return buf, nil
+}
+
+func (gcs *GoogleCloudStorageService) ExpirationPolicy(ctx context.Context) (daprovider.ExpirationPolicy, error) {
+	if gcs.enableExpiry {
+		return daprovider.KeepForever, nil
+	}
+	return daprovider.DiscardAfterDataTimeout, nil
+}
+
+func (gcs *GoogleCloudStorageService) Sync(ctx context.Context) error {
+	return nil
+}
+
+func (gcs *GoogleCloudStorageService) Close(ctx context.Context) error {
+	return gcs.operator.Close(ctx)
+}
+
+func (gcs *GoogleCloudStorageService) String() string {
+	return fmt.Sprintf("GoogleCloudStorageService(:%s)", gcs.bucket)
+}
+
+func (gcs *GoogleCloudStorageService) HealthCheck(ctx context.Context) error {
+	bucket := gcs.operator.Bucket(gcs.bucket)
+	// check if we have bucket permissions
+	permissions := []string{
+		"storage.buckets.get",
+		"storage.buckets.list",
+		"storage.objects.create",
+		"storage.objects.delete",
+		"storage.objects.list",
+		"storage.objects.get",
+	}
+	perms, err := bucket.IAM().TestPermissions(ctx, permissions)
+	if err != nil {
+		return fmt.Errorf("could not check permissions: %v", err)
+	}
+	sort.Strings(permissions)
+	sort.Strings(perms)
+	if !cmp.Equal(perms, permissions) {
+		return fmt.Errorf("permissions mismatch (-want +got):\n%s", cmp.Diff(permissions, perms))
+	}
+	// check if bucket exists (and others), and update expiration policy if enabled
+	attrs, err := bucket.Attrs(ctx)
+	if err != nil {
+		return err
+	}
+	if gcs.enableExpiry {
+		lifecycleRule := googlestorage.LifecycleRule{
+			Action:    googlestorage.LifecycleAction{Type: "Delete"},
+			Condition: googlestorage.LifecycleCondition{AgeInDays: int64(gcs.maxRetention.Hours() / 24)}, // Objects older than 30 days
+		}
+		attrs.Lifecycle.Rules = append(attrs.Lifecycle.Rules, lifecycleRule)
+
+		bucketAttrsToUpdate := googlestorage.BucketAttrsToUpdate{
+			Lifecycle: &attrs.Lifecycle,
+		}
+		if _, err := bucket.Update(ctx, bucketAttrsToUpdate); err != nil {
+			return fmt.Errorf("failed to update bucket lifecycle: %v", err)
+		}
+	}
```

Comment:
I think doing this on health check is not the right place. Can you do it on construction?

Reply:
ok

Reply:
Moved also.